contrib/gogearbox/gearbox.v1/gearbox.go

```diff
@@ -0,0 +1,91 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+// Package gearbox provides functions to trace the gogearbox/gearbox package (https://github.com/gogearbox/gearbox)
+package gearbox // import "gopkg.in/DataDog/dd-trace-go.v1/contrib/gogearbox/gearbox"
+
+import (
+	"fmt"
+	"strconv"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/contrib/gogearbox/gearbox.v1/internal/gearboxutil"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/telemetry"
+
+	"github.com/gogearbox/gearbox"
+	"github.com/valyala/fasthttp"
+)
+
+const componentName = "gogearbox/gearbox"
+
+func init() {
+	telemetry.LoadIntegration(componentName)
+}
+
+// Middleware returns middleware that will trace incoming requests.
+func Middleware(opts ...Option) func(gctx gearbox.Context) {
+	cfg := newConfig()
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	log.Debug("contrib/gogearbox/gearbox: Configuring Middleware: Service: %#v", cfg)
+	spanOpts := []tracer.StartSpanOption{
+		tracer.ServiceName(cfg.serviceName),
+	}
+	return func(gctx gearbox.Context) {
+		if cfg.ignoreRequest(gctx) {
+			gctx.Next()
+			return
+		}
+		fctx := gctx.Context()
+		spanOpts = defaultSpanTags(spanOpts, fctx)
+		// Create an instance of FasthttpCarrier, which embeds *fasthttp.RequestCtx and implements TextMapReader
+		fcc := &gearboxutil.FasthttpCarrier{
+			ReqHeader: &fctx.Request.Header,
+		}
+		if sctx, err := tracer.Extract(fcc); err == nil {
+			spanOpts = append(spanOpts, tracer.ChildOf(sctx))
+		}
+		span, _ := tracer.StartSpanFromContext(fctx, "http.request", spanOpts...)
+		defer span.Finish()
+
+		// AFAICT, there is no automatic way to update the fashttp context with the context returned from tracer.StartSpanFromContext
+		// Instead I had to manually add the activeSpanKey onto the fashttp context
+		activeSpanKey := tracer.ContextKey{}
+		fctx.SetUserValue(activeSpanKey, span)
+
+		gctx.Next()
+
+		span.SetTag(ext.ResourceName, cfg.resourceNamer(gctx))
+
+		status := fctx.Response.StatusCode()
+		if cfg.isStatusError(status) {
+			span.SetTag(ext.Error, fmt.Errorf("%d: %s", status, string(fctx.Response.Body())))
```

Comment:
Line 68 is the first instance in this PR where I'm doing byte-array-to-string conversion. I have to do this a lot, since fashttp methods return []byte types and Datadog/ ddtrace functionality expects a string.
[Fashttp readme](https://pkg.go.dev/github.com/valyala/fasthttp#section-readme) has a section called "Tricks with []byte buffers" where they expose `b2s` and `s2b` funcs that convert between the types without memory allocations. Fashttp encourages users to utilize the logic in their own code, but these functions use the unsafe package.

I'm curious if the frequent []byte <> string conversion is going to pose a risk to performance impact in this code and what some other alternatives are.